services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/northbound/switches/SwitchSpec.groovy

```diff
@@ -24,4 +33,118 @@ class SwitchSpec extends BaseSpecification {
+        lockKeeperService.knockoutSwitch(sw.getDpId())
+
+        when: "Try to delete inactive switch with ISLs"
+        northboundService.deleteSwitch(sw.getDpId(), false)
+
+        then: "Got 400 BadRequest because switch has ISLs"
+        def exc = thrown(HttpClientErrorException)
+        exc.rawStatusCode == 400
+
+        and: "Cleanup: Activate the switch back"
+        lockKeeperService.reviveSwitch(sw.getDpId())
+    }
+
+    def "Delete inactive switch with single flow"() {
+        requireProfiles("virtual")
+
+        given: "An active switch with flow"
+        def sw = topology.getActiveSwitches()[0]
+        def flow = flowHelper.singleSwitchFlow(sw)
+        northboundService.addFlow(flow)
+        assert Wrappers.wait(WAIT_OFFSET) { northboundService.getFlowStatus(flow.id).status == FlowState.UP }
+
+        and: "Deactivate this switch"
+        lockKeeperService.knockoutSwitch(sw.getDpId())
+
+        when: "Try to delete inactive switch with flow"
+        northboundService.deleteSwitch(sw.getDpId(), false)
+
+        then: "Got 400 BadRequest because switch has flow"
+        def exc = thrown(HttpClientErrorException)
+        exc.rawStatusCode == 400
+
+        and: "Cleanup: Activate the switch back and remove flow"
+        lockKeeperService.reviveSwitch(sw.getDpId())
+        northboundService.deleteFlow(flow.id)
+    }
+
+    def "Delete inactive switch with flow"() {
+        requireProfiles("virtual")
+
+        given: "An active switch with flow"
+        def (Switch srcSwitch, Switch dstSwitch) = topology.getActiveSwitches()[0..1]
+        def flow = flowHelper.randomFlow(srcSwitch, dstSwitch)
+        northboundService.addFlow(flow)
+        assert Wrappers.wait(WAIT_OFFSET) { northboundService.getFlowStatus(flow.id).status == FlowState.UP }
+
+        and: "Deactivate this switch"
+        lockKeeperService.knockoutSwitch(srcSwitch.getDpId())
+
+        when: "Try to delete inactive switch with flow"
+        northboundService.deleteSwitch(srcSwitch.getDpId(), false)
+
+        then: "Got 400 BadRequest because switch has flow"
+        def exc = thrown(HttpClientErrorException)
+        exc.rawStatusCode == 400
+
+        and: "Cleanup: Activate the switch back and remove flow"
+        lockKeeperService.reviveSwitch(srcSwitch.getDpId())
+        northboundService.deleteFlow(flow.id)
+    }
+
+    def "Force switch delete"() {
+        requireProfiles("virtual")
+
+        given: "An active switch"
+        def sw = topology.getActiveSwitches()[0]
+
+        when: "Try to force delete active switch with ISLs"
+        def response = northboundService.deleteSwitch(sw.getDpId(), true)
+
+        then: "Check that switch is actually deleted"
+        response.deleted
+        Wrappers.wait(WAIT_OFFSET) {
+            !northboundService.getAllSwitches().stream().anyMatch({ s -> (s.switchId == sw.dpId) })
+        }
+
+        and: "Cleanup: Restore the switch"
+        lockKeeperService.reviveSwitch(sw.getDpId())
```

Comment:
I'm wondering how this works. You didn't do `knockoutSwitch` in this test, so the switch theoretically is still connected to controller, so reviveSwitch should do nothing. Or does forced deleteSwitch operation also disconnects switch from controller?

Reply:
deleteSwitch just removes data from database. In real world switch should be disconnected from controller before delete. Force delete is just a way to delete switch without any restrictions. May be it will crash something in future. So it is not regular operation. Maybe reviveSwitch() do nothing on controller, but it restores the switch in database.  And it brings back topology on original state. This is the main purpose